charmcraft/commands/store/__init__.py

```diff
@@ -610,6 +621,10 @@ def _get_lib_info(*, full_name=None, lib_path=None):
             raise _BadLibraryNameError(full_name)
         lib_path = pathlib.Path('lib') / charmsdir / charm_name / v_api / (libfile + '.py')
 
+    # charm names in the path can contain '_' to be importable
+    # these should be '-', so change them back
+    charm_name = create_charm_name_from_importable(charm_name)
```

Comment:
My assumption here was that the fully qualified name would also be mutated to '_'. This would mean that full name would look like 'charms.test_charm.v0.testlib', and therefore needs transforming also.
Otherwise the fully qualified name is also an invalid import.

Reply:
Ah, you're right!!!!

It gets a little confusing.

Let me ask you to enhance the code with two things.

First, put the "rules" in `_get_lib_info`'s docstring, something like:

- `charm_name` (and `libdata.charm_name`): `foo-bar`
- `full_name` (and `libdata.full_name`): `charms.foo_bar.v0.somelib`
- any path (specially `libdata.path`): `lib/charms/foo_bar/v0/somelib`

Second, please rename `charm_name` to `importable_charm_name` where it corresponds. So it will be explicit that in the "if blocks" we're handling `importable_charm_name`, it won't be mixed afterwards, and even your conversion line is more straightforward:

    charm_name = create_charm_name_from_importable(importable_charm_name)